// @ng-web-apis/intersection-observer/esm5/directives/intersection-observee.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Inject } from '@angular/core';
import { Observable } from 'rxjs';
import { IntersectionObserveeService } from '../services/intersection-observee.service';
// @dynamic
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'rxjs';
var IntersectionObserveeDirective = /** @class */ (function () {
    function IntersectionObserveeDirective(waIntersectionObservee) {
        this.waIntersectionObservee = waIntersectionObservee;
    }
    /** @nocollapse */
    IntersectionObserveeDirective.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [IntersectionObserveeService,] }] }
    ]; };
IntersectionObserveeDirective.ɵfac = function IntersectionObserveeDirective_Factory(t) { return new (t || IntersectionObserveeDirective)(ɵngcc0.ɵɵdirectiveInject(IntersectionObserveeService)); };
IntersectionObserveeDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: IntersectionObserveeDirective, selectors: [["", "waIntersectionObservee", ""]], outputs: { waIntersectionObservee: "waIntersectionObservee" }, features: [ɵngcc0.ɵɵProvidersFeature([IntersectionObserveeService])] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(IntersectionObserveeDirective, [{
        type: Directive,
        args: [{
                selector: '[waIntersectionObservee]',
                outputs: ['waIntersectionObservee'],
                providers: [IntersectionObserveeService]
            }]
    }], function () { return [{ type: ɵngcc1.Observable, decorators: [{
                type: Inject,
                args: [IntersectionObserveeService]
            }] }]; }, null); })();
    return IntersectionObserveeDirective;
}());
export { IntersectionObserveeDirective };
if (false) {
    /** @type {?} */
    IntersectionObserveeDirective.prototype.waIntersectionObservee;
}
