// @ng-web-apis/intersection-observer/esm5/module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { IntersectionObserveeDirective } from './directives/intersection-observee.directive';
import { IntersectionObserverDirective } from './directives/intersection-observer.directive';
import { IntersectionRootDirective } from './directives/intersection-root.directive';
import * as ɵngcc0 from '@angular/core';
var IntersectionObserverModule = /** @class */ (function () {
    function IntersectionObserverModule() {
    }
IntersectionObserverModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: IntersectionObserverModule });
IntersectionObserverModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function IntersectionObserverModule_Factory(t) { return new (t || IntersectionObserverModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(IntersectionObserverModule, { declarations: function () { return [IntersectionObserverDirective,
        IntersectionObserveeDirective,
        IntersectionRootDirective]; }, exports: function () { return [IntersectionObserverDirective,
        IntersectionObserveeDirective,
        IntersectionRootDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(IntersectionObserverModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    IntersectionObserverDirective,
                    IntersectionObserveeDirective,
                    IntersectionRootDirective,
                ],
                exports: [
                    IntersectionObserverDirective,
                    IntersectionObserveeDirective,
                    IntersectionRootDirective,
                ]
            }]
    }], function () { return []; }, null); })();
    return IntersectionObserverModule;
}());
export { IntersectionObserverModule };
